crypto/src/signature.rs

```diff
@@ -125,6 +136,71 @@ impl ::std::fmt::Display for Signature {
     }
 }
 
+macro_rules! from_s_for_sig {
+    ($sig:ident, $name:ident) => {
+        impl From<$sig> for Signature {
+            fn from(s: $sig) -> Self {
+                Self::$name(s)
+            }
+        }
+    };
+}
+from_s_for_sig!(Ed25519Signature, Ed25519);
+from_s_for_sig!(Secp256k1Signature, Secp256k1);
+from_s_for_sig!(P256Signature, P256);
+from_s_for_sig!(BlsSignature, Bls);
+from_s_for_sig!(UnknownSignature, Unknown);
+
+#[derive(Debug, Error)]
+pub enum TryFromSignatureError {
+    #[error("Incorrect signature kind {0:?}.")]
+    InvalidKind(HashType),
+}
+
+macro_rules! from_sig_for_s {
+    ($sig:ident, $name:ident) => {
+        impl TryFrom<Signature> for $sig {
+            type Error = TryFromSignatureError;
+
+            fn try_from(s: Signature) -> Result<Self, Self::Error> {
+                match s {
+                    Signature::$name(s) => Ok(s),
+                    Signature::Unknown(s) => Ok(s.into()),
+                    s => Err(Self::Error::InvalidKind(s.hash_type())),
+                }
+            }
+        }
+    };
+}
+from_sig_for_s!(Ed25519Signature, Ed25519);
+from_sig_for_s!(Secp256k1Signature, Secp256k1);
+from_sig_for_s!(P256Signature, P256);
+
+impl TryFrom<Signature> for BlsSignature {
+    type Error = TryFromSignatureError;
+
+    fn try_from(s: Signature) -> Result<Self, Self::Error> {
+        match s {
+            Signature::Bls(s) => Ok(s),
+            s => Err(Self::Error::InvalidKind(s.hash_type())),
+        }
+    }
+}
+
+impl TryFrom<Signature> for UnknownSignature {
+    type Error = TryFromSignatureError;
+
+    fn try_from(s: Signature) -> Result<Self, Self::Error> {
+        match s {
+            Signature::Ed25519(s) => Ok(s.into()),
+            Signature::Secp256k1(s) => Ok(s.into()),
+            Signature::P256(s) => Ok(s.into()),
+            Signature::Unknown(s) => Ok(s),
+            s => Err(Self::Error::InvalidKind(s.hash_type())),
```

Comment:
BLS doesn't work?

Reply:
Yeah, see here in [octez](https://gitlab.com/tezos/tezos/-/blob/master/src/lib_crypto/signature_v1.ml?ref_type=heads#L571)

TLDR; the signature length for all signatures is 64 bytes, *except for bls*, which is 96. Therefore everything can be a valid `unknown`, except for BLS sigs

Reply:
Ok, I understand. But this is very confusing from a UX perspective. E.g. why does an unknown signature have a fixed length? I think this abstraction needs to be re-thought.

Reply:
I do kind of agree - it's definitely worth thinking about for future, especially if we can improve things on the octez side too.